### PR TITLE
Move bn to pten

### PR DESCRIPTION
**Pull request by phlrain**
### PR types
 Breaking changes 

### PR changes
OPs

### Describe

move batch norm to pten

### CODE REVIEW

File: paddle/fluid/framework/operator.cc [hunk 1]
```diff
@@ -1906,7 +1905,12 @@ Scope* OperatorWithKernel::PreparePtenData(
 
   for (size_t i = 0; i < input_defs.size(); ++i) {
     auto& in_def = input_defs.at(i);
-    auto& ins_vector = ctx->inputs.at(input_names[i]);
+    auto it = ctx->inputs.find(input_names[i]);
+    if (it == ctx->inputs.end()) {
+      continue;
+    }
+    // auto& ins_vector = ctx->inputs.at(input_names[i]);
```
Comment (chenwhql) [inline paddle/fluid/framework/operator.cc]: 注释无用的话，建议删除

File: paddle/fluid/framework/operator.cc [hunk 2]
```diff
@@ -1906,7 +1905,12 @@ Scope* OperatorWithKernel::PreparePtenData(
 
   for (size_t i = 0; i < input_defs.size(); ++i) {
     auto& in_def = input_defs.at(i);
-    auto& ins_vector = ctx->inputs.at(input_names[i]);
+    auto it = ctx->inputs.find(input_names[i]);
```
Comment (chenwhql) [inline paddle/fluid/framework/operator.cc]: name找不到的情况是怎么产生的？

File: paddle/pten/kernels/batch_norm_grad_grad_kernel.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+#pragma once
+
+#include <string>
+#include "paddle/pten/core/dense_tensor.h"
+
+namespace pten {
+
+template <typename T, typename Context>
+void BatchNormGradGradKernel(const Context& dev_ctx,
```
Comment (chenwhql) [inline paddle/pten/kernels/batch_norm_grad_grad_kernel.h]: 命名风格要不要统一下，DoubleGrad？

File: paddle/pten/kernels/cpu/batch_norm_grad_grad_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+  PADDLE_ENFORCE_EQ(is_test,
+                    false,
+                    pten::errors::InvalidArgument(
+                        "`is_test = True` CANNOT be used in train program. If "
+                        "you want to use global status in pre_train model, "
+                        "please set `use_global_stats = True`"));
+
+  const auto data_layout =
+      paddle::framework::StringToDataLayout(data_layout_str);
```
Comment (chenwhql) [inline paddle/pten/kernels/cpu/batch_norm_grad_grad_kernel.cc]: [TODO ] 这个我后面也移动到pten

File: paddle/pten/kernels/cpu/batch_norm_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+    EigenVectorArrayMap<T> running_mean_arr(
+        mean_out->mutable_data<T>(ctx.GetPlace()), C);
+    EigenVectorArrayMap<T> running_var_arr(
+        variance_out->mutable_data<T>(ctx.GetPlace()), C);
+
+    if ((N * sample_size) == 1) {
+      // Only 1 element in normalization dimension,
+      // we skip the batch norm calculation, let y = x.
+      paddle::framework::TensorCopy(x, ctx.GetPlace(), y);
```
Comment (chenwhql) [inline paddle/pten/kernels/cpu/batch_norm_kernel.cc]: 这里可以复用pten::CopyKernel吗

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 6]
```diff
@@ -0,0 +1,979 @@
...
+#define LAUNCH_BOUNDS(BlockDim) __launch_bounds__(BlockDim)
+#else
+#define LAUNCH_BOUNDS(BlockDim)
+#endif
+
+DECLARE_bool(cudnn_batchnorm_spatial_persistent);
+namespace pten {
+
+template <typename T>
+using CudnnDataType = paddle::platform::CudnnDataType<T>;
```
Comment (chenwhql) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: [TODO] 我后面移动一下

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 7]
```diff
@@ -0,0 +1,979 @@
...
+template <typename T>
+using BatchNormParamType = typename CudnnDataType<T>::BatchNormParamType;
+
+template <typename T, int BlockDim, paddle::framework::DataLayout layout>
```
Comment (chenwhql) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: pten::DataLayout

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 8]
```diff
@@ -0,0 +1,979 @@
...
+    db_sum = BlockReduce(db_storage).Reduce(db_sum, cub::Sum());
+    if (threadIdx.x == 0) {
+      dscale[i] = ds_sum * inv_var_i;
+      dbias[i] = db_sum;
+    }
+    __syncthreads();
+  }
+}
+
+template <typename T, paddle::framework::DataLayout layout>
```
Comment (chenwhql) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: 同上

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 9]
```diff
@@ -0,0 +1,979 @@
...
+        layout == paddle::framework::DataLayout::kNCHW ? i / HxW % C : i % C;
+    BatchNormParamType<T> inv_var = 1.0 / sqrt(variance[c] + epsilon);
+    dx[i] = static_cast<T>(static_cast<BatchNormParamType<T>>(dy[i]) *
+                           scale[c] * inv_var);
+  }
+}
+
+template <typename T>
+static __global__ void KeBNRestoreData(
+    const paddle::framework::DataLayout layout,
```
Comment (chenwhql) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: 同上，还有几处，建议文件内搜索替换一下

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 10]
```diff
@@ -0,0 +1,979 @@
...
+  auto *d_scale = scale_grad;
+  auto *d_bias = bias_grad;
+
+  // batch_norm with inplace as false will take X as grad input, which
+  // is same as cuDNN batch_norm backward calculation, batch_norm
+  // with inplace as true only take Y as input and X should be calculate
+  // by inverse operation of batch_norm on Y
+  // const DenseTensor *x;
+
+  // if (ctx.HasInput("Y")) {
```
Comment (chenwhql) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: 这段注释是可以清理的吗

File: python/paddle/fluid/tests/unittests/dygraph_to_static/test_mobile_net.py [hunk 11]
```diff
@@ -616,6 +617,7 @@ def test_mobile_net(self):
         self.assert_same_loss("MobileNetV1")
         # MobileNet-V2
         self.assert_same_loss("MobileNetV2")
+        print("assert mobile fin")
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/dygraph_to_static/test_mobile_net.py]: 调试用的？

File: python/paddle/fluid/tests/unittests/test_batch_norm_op.py [hunk 12]
```diff
@@ -496,6 +496,7 @@ def test_with_place(place, data_layout, shape):
 
                 program._sync_with_cpp()
 
+                #print( program )
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_batch_norm_op.py]: 可以移除？

File: python/paddle/fluid/tests/unittests/test_batch_norm_op.py [hunk 13]
```diff
@@ -716,4 +717,6 @@ def test_reservespace(self):
 
 
 if __name__ == '__main__':
+    import paddle
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_batch_norm_op.py:719]: import paddle建议放到最前面

File: python/paddle/fluid/tests/unittests/test_batch_norm_op_v2.py [hunk 14]
```diff
@@ -300,4 +300,6 @@ def init_test(self):
 
 
 if __name__ == '__main__':
+    import paddle
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_batch_norm_op_v2.py:303]: 同上

File: paddle/pten/kernels/batch_norm_grad_grad_kernel.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline paddle/pten/kernels/batch_norm_grad_grad_kernel.h]: 目前反向一二三阶的反向kernel都放在一个文件里，这样能减少一些文件数量

File: paddle/pten/kernels/gpu/batch_norm_grad_grad_kernel.cu [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+  const DataLayout data_layout =
+      paddle::framework::StringToDataLayout(data_layout_str);
+
+  const DenseTensor* running_mean = nullptr;
+  const DenseTensor* running_variance = nullptr;
+  if (use_global_stats) {
+    running_mean = mean.get_ptr();
+    running_variance = variance.get_ptr();
+  }
+  paddle::operators::NormDoubleGradFunctor<Context, T>(ctx,
```
Comment (zyfncg) [inline paddle/pten/kernels/gpu/batch_norm_grad_grad_kernel.cu]: 这个`NormDoubleGradFunctor`是不也可以移过来？

File: paddle/pten/kernels/gpu/batch_norm_utils.h [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "paddle/fluid/framework/op_registry.h"
```
Comment (zyfncg) [inline paddle/pten/kernels/gpu/batch_norm_utils.h]: 这个op_registry.h的头文件可以去掉吗？

File: paddle/pten/kernels/batch_norm_grad_grad_kernel.h (hunk 3 above)
Comment (phlrain) [reply paddle/pten/kernels/batch_norm_grad_grad_kernel.h]: 可以，我改下

File: paddle/fluid/framework/operator.cc (hunk 2 above)
Comment (phlrain) [reply paddle/fluid/framework/operator.cc]: 有些GradOpMaker在实现的时候会判断HasInput，然后在SetInput，这个input有时候就会没有

File: paddle/pten/kernels/gpu/batch_norm_grad_grad_kernel.cu (hunk 16 above)
Comment (phlrain) [reply paddle/pten/kernels/gpu/batch_norm_grad_grad_kernel.cu]: 可以，我在单独拉一个pr，挪过来吧。。这个pr有点大了，我先保证功能正确，再把其他的组件移动过来

File: paddle/pten/kernels/batch_norm_grad_grad_kernel.h (hunk 15 above)
Comment (phlrain) [reply paddle/pten/kernels/batch_norm_grad_grad_kernel.h]: 好的，我整理一下。

File: paddle/pten/kernels/gpu/batch_norm_utils.h (hunk 17 above)
Comment (phlrain) [reply paddle/pten/kernels/gpu/batch_norm_utils.h]: 可以去，我删一下

File: paddle/fluid/operators/norm_utils.cu.h [hunk 18]
```diff
@@ -389,11 +389,12 @@ __global__ void DoubleGradComputeDDYWithGlobal(
 }
 
 template <typename DeviceContext, typename T>
-void NormDoubleGradFunctor(const framework::ExecutionContext &ctx,
+void NormDoubleGradFunctor(const DeviceContext &ctx,
```
Comment (ceci3) [inline paddle/fluid/operators/norm_utils.cu.h:392]: 这个计算不需要放到pten里面嘛？

File: paddle/pten/ops/compat/batch_norm_sig.cc [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+       "epsilon",
+       "data_layout",
+       "is_test",
+       "use_global_stats",
+       "trainable_statistics",
+       "fuse_with_relu"},
+      {GradVarName("X"), GradVarName("Scale"), GradVarName("Bias")});
+}
+
+KernelSignature BatchNormGradGradOpArgumentMapping(
```
Comment (ceci3) [inline paddle/pten/ops/compat/batch_norm_sig.cc]: GradGrad -> DoubleGrad？

File: paddle/fluid/operators/inplace_abn_op.cc [hunk 20]
```diff
@@ -207,7 +208,33 @@ class InplaceABNKernel
     auto activation =
         GetInplaceABNActivationType(ctx.Attr<std::string>("activation"));
     auto& place = *ctx.template device_context<DeviceContext>().eigen_device();
-    BatchNormKernel<DeviceContext, T>::Compute(ctx);
+
```
Comment (ceci3) [inline paddle/fluid/operators/inplace_abn_op.cc:217]: 这里是把inplace abn的逻辑挪出来了吗？

File: paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu [hunk 21]
```diff
@@ -0,0 +1,979 @@
...
+      x_dims.size() >= 2 && x_dims.size() <= 5,
+      true,
+      pten::errors::InvalidArgument(
+          "The size of input's dimensions should be between 2 and 5."
+          "But received: the size of input's dimensions is [%d],"
+          "the dimensions of input is [%s]",
+          x_dims.size(),
+          x_dims));
+  int N, C, H, W, D;
+  paddle::operators::ExtractNCWHD(x_dims, data_layout, &N, &C, &H, &W, &D);
```
Comment (ceci3) [inline paddle/pten/kernels/gpu/batch_norm_grad_kernel.cu]: 这个函数还要放在operator下面嘛？

File: paddle/phi/kernels/cpu/batch_norm_grad_kernel.cc [hunk 22]
```diff
@@ -0,0 +1,681 @@
...
+    paddle::framework::TensorCopy(*d_y, ctx.GetPlace(), d_x);
+    return;
+  }
+
+  int scale_coefff = use_global_stats ? 1 : N * sample_size;
+  const auto scale_inv_var_nhw = scale_arr * inv_var_arr / scale_coefff;
+
+  DenseTensor dy_sum;
+  dy_sum.Resize({C});
+  dy_sum.mutable_data<T>(ctx.GetPlace());
```
Comment (chenwhql) [inline paddle/phi/kernels/cpu/batch_norm_grad_kernel.cc]: 好像还有一些mutable_data，需要替换成ctx.Alloc，建议全局搜索替换一下

File: paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu [hunk 23]
```diff
@@ -0,0 +1,1039 @@
...
+    kernel->OutputAt(2).SetDataType(phi::DataType::FLOAT32);
+    kernel->OutputAt(3).SetDataType(phi::DataType::FLOAT32);
+    kernel->OutputAt(4).SetDataType(phi::DataType::FLOAT32);
+  }
+}
+
+#endif
+
+#ifdef PADDLE_WITH_HIP
+PD_REGISTER_KERNEL(batch_norm_grad_grad,
```
Comment (chenwhql) [inline paddle/phi/kernels/gpu/batch_norm_grad_kernel.cu:1024]: 这个命名我们后续需要统一一下写法

File: paddle/phi/kernels/gpu/batch_norm_kernel.cu [hunk 24]
```diff
@@ -0,0 +1,680 @@
...
+  const auto &x_dims = x.dims();
+  PADDLE_ENFORCE_EQ(
+      x_dims.size() >= 2 && x_dims.size() <= 5,
+      true,
+      phi::errors::InvalidArgument(
+          "The size of input's dimensions should be between 2 and 5"
+          "But received: the size of input's dimensions is [%d]",
+          x_dims.size()));
+
+  y->mutable_data<T>(ctx.GetPlace());
```
Comment (chenwhql) [inline paddle/phi/kernels/gpu/batch_norm_kernel.cu]: mutable_data -> Alloc

**Review by chenwhql — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
